router/h2/src/main/scala/io/buoyant/router/h2/StreamStatsFilter.scala

```diff
@@ -1,83 +1,130 @@
   val role = Stack.Role("StreamStatsFilter")
   val module: Stackable[ServiceFactory[Request, Response]] =
-    new Stack.Module1[param.Stats, ServiceFactory[Request, Response]] {
+    new Stack.Module2[param.Stats, h2param.H2ResponseClassifier, ServiceFactory[Request, Response]] {
       override def role = StreamStatsFilter.role
       override def description = "Record stats on h2 streams"
-      override def make(statsP: param.Stats, next: ServiceFactory[Request, Response]) = {
+      override def make(
+        statsP: param.Stats,
+        classifierP: h2param.H2ResponseClassifier,
+        next: ServiceFactory[Request, Response]
+      ) = {
         val param.Stats(stats) = statsP
-        new StreamStatsFilter(stats).andThen(next)
+        val h2param.H2ResponseClassifier(classifier) = classifierP
+        new StreamStatsFilter(stats, classifier).andThen(next)
       }
     }
 }
 
-class StreamStatsFilter(statsReceiver: StatsReceiver)
+class StreamStatsFilter(statsReceiver: StatsReceiver, classifier: H2ResponseClassifier)
   extends SimpleFilter[Request, Response] {
 
-  class StreamStats(
-    stats: StatsReceiver,
-    durationName: Option[String] = None,
-    successName: Option[String] = None
-  ) {
+  class StreamStats(protected val stats: StatsReceiver, durationName: Option[String] = None) {
 
     private[this] val durationMs =
       stats.stat(s"${durationName.getOrElse("stream_duration")}_ms")
-    private[this] val successes =
-      stats.counter(s"${successName.getOrElse("stream_")}successes")
-    private[this] val failures =
-      stats.counter(s"${successName.getOrElse("stream_")}failures")
+    private[this] val successes = stats.counter("stream_successes")
+    private[this] val failures = stats.counter("stream_failures")
 
-    @inline def apply(startT: Stopwatch.Elapsed)(result: Try[_]): Unit = {
+    @inline def onEnd(startT: Stopwatch.Elapsed)(result: Try[_]): Unit = {
       durationMs.add(startT().inMillis)
       result match {
         case Return(_) => successes.incr()
         case Throw(_) => failures.incr()
       }
     }
+
+    @inline def apply(startT: Stopwatch.Elapsed)(result: Try[_]): Unit = onEnd(startT)(result)
+
+  }
+
+  trait FrameStats extends StreamStats {
+    private[this] val frameSize = stats.stat("data_frame", "bytes")
+    private[this] val frameCount = stats.stat("data_frame", "count")
+
+    def onFrame(startT: Stopwatch.Elapsed)(stream0: Stream): Stream = {
+      var streamFrameCount = 0
+      val stream1 = stream0.onFrame {
+        case Return(frame: Frame.Data) =>
+          frameSize.add(frame.buf.length)
+          streamFrameCount += 1
+        case _ =>
+      }
+      val _ = stream1.onEnd.respond { result =>
+        frameCount.add(streamFrameCount)
+        this.onEnd(startT)(result)
+      }
+      stream1
+    }
+
+    @inline def apply(startT: Stopwatch.Elapsed)(stream: Stream): Stream = onFrame(startT)(stream)
+  }
+
+  object ClassifierStats {
```

Comment:
This class level object is pretty confusing.  I think we can get rid of it and move its contents up into the class directly.